### DISCUSSION TITLE
how can I override foreign key constraint name?

### DISCUSSION THREAD

**Discussion by SDAdham** [mikro-orm/mikro-orm#1202] [Q&A]
how can I override foreign key constraint name? I lookedup the code and it looks like there is no a specific config that allows me to override it...

The reason behind this request is because for some columns I am getting foreign key index name is too long. Please help about this. It's urgent.

A more detailed context of why I'ma sking for this:
```
UnhandledPromiseRejectionWarning: DriverException: alter table `table_name` add index `table_name_column_name_id_index`(`column_name_id`); - Identifier name 'table_name_column_name_id_index' is too long
    at MySqlExceptionConverter.convertException (\node_modules\@mikro-orm\core\platforms\ExceptionConverter.js:8:16)
    at MySqlExceptionConverter.convertException (\node_modules\@mikro-orm\mysql-base\MySqlExceptionConverter.js:78:22)
    at MySqlDriver.convertException (\node_modules\@mikro-orm\core\drivers\DatabaseDriver.js:171:54)
    at \node_modules\@mikro-orm\core\drivers\DatabaseDriver.js:175:24
    at ZoneDelegate.invoke (\node_modules\zone.js\dist\zone-node.js:386:30)
    at Zone.run (\node_modules\zone.js\dist\zone-node.js:143:47)
    at \node_modules\zone.js\dist\zone-node.js:891:38
    at ZoneDelegate.invokeTask (\node_modules\zone.js\dist\zone-node.js:421:35)
    at Zone.runTask (\node_modules\zone.js\dist\zone-node.js:188:51)
    at drainMicroTaskQueue (\node_modules\zone.js\dist\zone-node.js:601:39)

previous Error: alter table `table_name` add index `table_name_column_name_id_index`(`column_name_id`); - Identifier name 'table_name_column_name_id_index' is too long
    at Packet.asError (\node_modules\mysql2\lib\packets\packet.js:712:17)
    at Query.execute (\node_modules\mysql2\lib\commands\command.js:28:26)
    at Connection.handlePacket (\node_modules\mysql2\lib\connection.js:425:32)
    at PacketParser.onPacket (\node_modules\mysql2\lib\connection.js:75:12)
    at PacketParser.executeStart (\node_modules\mysql2\lib\packet_parser.js:75:16)
    at Socket.<anonymous> (\node_modules\mysql2\lib\connection.js:82:25)
    at ZoneDelegate.invokeTask (\node_modules\zone.js\dist\zone-node.js:421:35)
    at Zone.runTask (\node_modules\zone.js\dist\zone-node.js:188:51)
    at ZoneTask.invokeTask (\node_modules\zone.js\dist\zone-node.js:503:38)
    at Socket.ZoneTask.invoke (\node_modules\zone.js\dist\zone-node.js:492:52)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:22956) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
```

**Comment by alexojegu** [comment]
Maybe you are referring to `joinColumn` or `inverseJoinColumn`:
  - https://mikro-orm.io/docs/decorators#onetoone
  - https://mikro-orm.io/docs/decorators#onetomany
  - https://mikro-orm.io/docs/decorators#manytomany

Or to naming strategy:
  - https://mikro-orm.io/docs/naming-strategy/

**Reply by SDAdham** [reply]
Hello xeladejo, I did use joinColumn instead and changed the join column name to a very minimal column name which was not favorable. and it wasn't really encouraged nor recommended. :(

**Comment by B4nan** [comment]
```ts
@ManyToOne({ index: 'custom_idx_name' })
```

**Reply by alexojegu** [reply]
That decorator parameter seems undocumented?
  - https://mikro-orm.io/docs/decorators/#manytoone

**Reply by SDAdham** [reply]
Hello B4nan & xeladejo, I used this decorator, and it doesn't help.

I wish it would help, but no, see this code https://github.com/mikro-orm/mikro-orm/blob/742b8131ba7d0acec2cca3f289237fd0e757baa5/packages/postgresql/src/PostgreSqlConnection.ts#L69 where it only utilizes table name and column name.

**Reply by B4nan** [reply]
That code is unrelated to your problem, its postgres driver (and it is about enum column diffing), you said you are using mysql and your issue is FK index name. 

The code I sent you is how you can adjust the index name. Show us your entity definition if it does not work for you, just saying so won't help with anything.

**Reply by SDAdham** [reply]
Hello b4nan,

i'm not sure you already tried the solution you provided or you are just guessing based on the name, but tbh, i really tried it and yet still get the same error, i even deleted `temp` and `dist` folders and restarted the entire build with a blank DB and yet the error still happens, you can try it using an long class names, take the one below as example:

```
export class AssassinsCreedOdyseeGamerTagSetting {
@ManyToOne(() => UserHaveAssassinsCreed, {
    nullable: true,
    index: 'acogts_uhac_index',
  })
  userHaveAssassinsCreed: UserHaveAssassinsCreed;
}
```

And also, this is not happening with just 1 class, any class with long name or a long column.

**Reply by SDAdham** [reply]
another note, the error that comes is not about `acogts_uhac_index`, it's on the constraint name `assassins_creed_odysee_gamer_tag_setting_user_have_assassins_creed_id_index`. So `index` i'm assuming has a different purpose.

**Reply by B4nan** [reply]
The error you provided in the OP is about index: `Error: alter table `table_name` add index `table_name_column_name_id_index`(`column_name_id`); - Identifier name 'table_name_column_name_id_index' is too long`. 

> i'm not sure you already tried the solution you provided or you are just guessing based on the name

It sounds more like you are the one who is guessing as your stack trace is about mysql and you are trying to tell me the postgres driver has anything to do with it? :] Guess what, I wrote most of this library, so yes, I tried...

> another note, the error that comes is not about acogts_uhac_index, it's on the constraint name assassins_creed_odysee_gamer_tag_setting_user_have_assassins_creed_id_index. So index i'm assuming has a different purpose.

If the issue is with FK constraint, then your OP is totally misleading, as the error you provided is not about FK, but about regular index (yes, mysql requires you to have indexes on FKs). 

The FK constraint names might not be currently configurable (its not MikroORM who is computing that name, it's knex). But again, if that is your issue, you are providing different stack trace than you are talking about.

**Reply by SDAdham** [reply]
Hello B4nan, I just retried now, and here another error:
```
alter table `table_name` add constraint `table_name_column_name_id_foreign` foreign key (`column_name_id`) references `table_name` (`id`) on update cascade on delete set null; - Identifier name 'table_name_column_name_id_foreign' is too long
```

You are right. I just realized that i was experiencing both not just 1 issue (maybe that what was confusing me and made me feel that `index` is not solving anything), i just didn't notice the difference because the error are almost the same in writing and the difference is from `index` to `constraint`.

**Comment by Wedrix** [comment]
Hello, I'm also facing the same problem: the constraint name is too long. @B4nan is there really no other way to solve this apart from renaming or disabling the foreign key constraint?

**Comment by Wedrix** [comment]
@SDAdham my workaround was to create a custom naming strategy that extends `UnderscoreNamingConvention` and override the `joinKeyColumnName` method:

```
~ mikro-orm.config.ts



import { UnderscoreNamingStrategy } from "@mikro-orm/core";

class CustomNamingStrategy extends UnderscoreNamingStrategy {

    joinKeyColumnName(entityName: string, referencedColumnName: string): string {
        return entityName.substr(0, 4) + (referencedColumnName || this.referenceColumnName()).substr(0, 4);
    }

};

import config from "./config";

export default {
    ....
    namingStrategy: CustomNamingStrategy,
};

```

**Comment by SDAdham** [comment]
Thanks @Wedrix

**Comment by B4nan** [comment]
FYI the issue with too long index/fk names will be addressed via #1297 (they will get automatically truncated to 64 chars).

edit: merged, will be available as `4.3.5-dev.69`

**Reply by Wedrix** [reply]
@B4nan thanks for the great work! Will it be possible to use something guaranteed to be unique like a UUID instead? 
It seems Doctrine handles it this way, prefixing the keys with IDX_ and the constraints with FK_.

**Reply by Wedrix** [reply]
@B4nan I just reviewed the PR. It should work. Thanks!
